src/MongoDB.Driver/Linq/Processors/ExpressionValidator.cs

```diff
@@ -0,0 +1,122 @@
+using System.Linq;
+using System.Linq.Expressions;
+
+namespace MongoDB.Driver.Linq.Processors
+{
+    /// <summary>
+    /// The expression visitor to validate a LINQ expression.
+    /// </summary>
+    internal sealed class ExpressionValidator : ExpressionVisitor
+    {
+        private readonly List<string> _topLevelInParameters = new List<string>();
+        private List<string> _inParametersScope = new List<string>();
+
+        /// <summary>
+        /// Validates the input expression.
+        /// </summary>
+        /// <param name="expression">The expression.</param>
+        public static void Validate(Expression expression)
+        {
+            var validator = new ExpressionValidator();
+
+            var lambdaExpression = expression as LambdaExpression;
+            if (lambdaExpression?.Parameters != null && lambdaExpression.Parameters.Any())
+            {
+                validator._topLevelInParameters.AddRange(lambdaExpression.Parameters.Select(c => c.Name));
+            }
+
+            validator.Visit(expression);
+        }
+
+        /// <inheritdoc />
+        protected override Expression VisitMethodCall(MethodCallExpression node)
+        {
+            if (node.Arguments.Any())
+            {
+                foreach (var argument in node.Arguments.Where(c => !(c is ConstantExpression || c is ParameterExpression))) // do not check the bottom of tree branches.
+                {
+                    var usedParameters = ParameterExpressionCollector.Collect(argument).ToList();
+
+                    if (!usedParameters.Any())
+                    {
+                        continue;
+                    }
+
+                    var memberExpression = argument as MemberExpression;
+                    if (memberExpression != null)
+                    {
+                        // add top-level expression parameters to the list
+                        // with parent's parameters.
+                        // NOTE: these parameters will be added to all members,
+                        // even if the member is not in the top-level, but since
+                        // top-level parameters should not be placed in any level 
+                        // except of direct children, it's not a problem
+                        AddTopLevelExpressionParameters(_inParametersScope);
+                        continue;
+                    }
+
+                    foreach (var usedParameter in usedParameters)
+                    {
+                        // check the current LINQ expression on whether
+                        // the parent expression parameters are used
+                        // on the current expression level or not.
+                        if (_inParametersScope.Contains(usedParameter))
+                        {
+                            throw new NotSupportedException($"The LINQ expression: {argument} has the member \"{usedParameter}\" which can not be used to build a correct Mongo query");
```

Comment:
I assume we're attempting to detect use of parameters that are in scope in C# but that can't be translated to a MongoDB field reference?

I would have expected a test more like:

    if (!validParameters.Contains(usedParameter))
    {
        throw ...
    }

So I'm confused why we're using Contains instead of !Contains.

Reply:
>I assume we're attempting to detect use of parameters that are in scope in C# but that can't be translated to a MongoDB field reference?

yep, you are right.
>I would have expected a test more like:

the idea of this test is if the current expression level use parameters which were defined in higher levels => throwing exception. So, we don't have list of `valid` parameters, we only have the list with `"invalid"`.

Reply:
This seems backwards. Why would we have a list of "invalid" parameters (which would have to be collected from multiple levels above us) instead of just having a list of "valid" parameters.

Reply:
Because valid parameters are all parameters except `invalid` ..
Can you advice how we can have a list of `valid` parameters?

Reply:
I would have expected that there is only ONE parameter valid at any point in time. For example, in:
 
    ...Where(x => expression)

The ONLY parameter reference valid in expression should be "x".

Reply:
`Expression<Func<TestObject, bool>> expr = (a) => local.Any(b => a.Collection2.Contains(b))`
This query is valid.

Reply:
OK. Good example.

So why can't we tell from that the valid parameters initially are "a" and then inside Any "b" becomes valid also.

Reply:
I guess it also can be a solution here. But I'm not sure that it will be easier.
At least we will need to change the logic of `ParameterExpressionCollector` to collect only parameters from the current level (instead of the current + children as it implemented right now).
Also, we will need to check the content of the current node (at least is there a local collection or not).
I think it's not a full list of possible changes in this case.